snare

```diff
@@ -0,0 +1,68 @@
+#! /usr/bin/env python3.7
+
+# Copyright (c) 2019 King's College London, created by the Software Development
+# Team <http://soft-dev.org/>
+#
+# Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+# http://www.apache.org/licenses/LICENSE-2.0>, or the MIT license <LICENSE-MIT
+# or http://opensource.org/licenses/MIT>, or the UPL-1.0 license
+# <http://opensource.org/licenses/UPL> at your option. This file may not be
+# copied, modified, or distributed except according to those terms.
+
+import getopt, hashlib, hmac, json, os, subprocess, sys, tempfile, traceback
+from http.server import ThreadingHTTPServer, BaseHTTPRequestHandler
+from urllib.parse import unquote
+
+class HookHandler(BaseHTTPRequestHandler):
+    def do_POST(self):
+        data = self.rfile.read(int(self.headers["Content-Length"]))
+        if not self._check_secret(data):
+            self.send_response(401)
+            return
+
+        data_str = unquote(data.decode("utf-8")[8:])
+        pl = json.loads(data_str)
+        owner = pl["repository"]["owner"]["login"].lower()
+        repo = pl["repository"]["name"].lower()
+
+        path = os.path.join(repo_scripts_dir, "%s_%s" % (owner, repo))
+        if os.path.exists(path):
+            with tempfile.NamedTemporaryFile() as jsonf:
+                jsonf.write(data_str.encode())
+                event = self.headers["X-GitHub-Event"]
+                subprocess.run([path, event, jsonf.name])
+
+        self.send_response(200)
```

Comment:
We may want to allow users to return their own status codes and messages here. For example, they may want to return something like `400, "Not yet implemented"`, which will then show up in their GitHub log (e.g. https://github.com/<user>/settings/hooks/<hookid>), making it easier to debug their scripts.

Reply:
I'm not really sure what to do with the return codes because there's no guarantee that the script completes in time to return a code (e.g. yksom's takes long enough to complete that GitHub assumes that things have timed out). That said, returning 400 if there isn't a script is probably a good idea

Reply:
See if https://github.com/softdevteam/snare/pull/1/commits/78a3e3ad97a0973fcb76819e3f9324f6ab5875d5 makes things a bit better. It's not configurable, but it's a little better I think.

Reply:
> there's no guarantee that the script completes in time to return a code

That's fine, in those cases we are already not returning `200`, but in the cases we do see a return code I think it would still be nice if users could define it themselves along with a message. When I did the plugin for Mattermost it was very useful to see which hooks I haven't implemented yet as they would be highlighted in the GitHub log.

Reply:
Maybe we can do customisation in the future? I'm not sure how to handle it sensibly at the moment, and we don't really need it at the moment either.

Reply:
I'm happy to make this a future feature, I just wanted to raise it as from my own experience working with GitHub webhooks this can be very useful.